codalab/server/bundle_manager.py

```diff
@@ -592,12 +595,11 @@ def _schedule_run_bundles(self):
         self._cleanup_dead_workers(workers)
         self._restage_stuck_starting_bundles(workers)
         self._bring_offline_stuck_running_bundles(workers)
-        self._fail_on_too_many_resources()
         self._acknowledge_recently_finished_bundles(workers)
+        staged_bundles_to_run = self._fail_on_too_many_resources()
 
         # Schedule, preferring user-owned workers.
-        self._schedule_run_bundles_on_workers(workers, user_owned=True)
-        self._schedule_run_bundles_on_workers(workers, user_owned=False)
+        self._schedule_run_bundles_on_workers(workers, staged_bundles_to_run)
```

Comment:
I like the idea of getting the bundles once and caching them, but maybe there's room for even more improvement here. Right now (ie before this PR) every operation here re-fetches everything from the database before trying to operate on them, but this doesn't really prevent race conditions and makes us waste a bunch of time re-getting everything from the database.

I wonder if we can juts get everythin (workers, bundles etc) once per bundle manager iteration, and explicitly skip cases where we notice  some disparity, hoping that next iteration they will be handled fine.

Reply:
That's true. We need to cache more variables so that we don't need to keep fetching them from the database.